File: src/temp/primitives/types/PDouble.kt
package temp.primitives.types

import temp.primitives.Primitive

class PDouble(var value: Double) : Primitive() {
    override fun getPrimitiveValue(): Any {
        return this.value
    }

    override fun negate(): Primitive {
        return PDouble(-this.value)
    }

    override fun not(): Primitive {
        return PNone()
    }

    override fun plus(primitive: Primitive): Primitive {
        return PDouble(this.value + (primitive as PDouble).value)
    }

    override fun minus(primitive: Primitive): Primitive {
        return PDouble(this.value - (primitive as PDouble).value)
    }

    override fun times(primitive: Primitive): Primitive {
        return PDouble(this.value * (primitive as PDouble).value)
    }

    override fun div(primitive: Primitive): Primitive {
        return PDouble(this.value / (primitive as PDouble).value)
    }

    override fun rem(primitive: Primitive): Primitive {
        return PDouble(this.value % (primitive as PDouble).value)
    }

    override fun compareGreater(primitive: Primitive): Primitive {
        return PBoolean(this.value > (primitive as PDouble).value)
    }

    override fun compareLesser(primitive: Primitive): Primitive {
        return PBoolean(this.value < (primitive as PDouble).value)
    }

    override fun toString(): String {
        return value.toString()
    }
}

File: src/codepulsar/primitive/type/PDouble.cpp
#include "PDouble.h"


Pulsar::PDouble::PDouble(double value) {
    this->value = value;
}

Pulsar::PrimitiveType Pulsar::PDouble::getPrimitiveType() {
    return PR_DOUBLE;
}

bool Pulsar::PDouble::isPrimitiveType(PrimitiveType primitiveType) {
    return primitiveType == PR_DOUBLE;
}

std::any Pulsar::PDouble::getPrimitiveValue() {
    return this->value;
}

Pulsar::Primitive* Pulsar::PDouble::unaryNegate() {
    return new PDouble(-this->value);
}

Pulsar::Primitive* Pulsar::PDouble::unaryNot() {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PDouble::plus(Primitive* primitive) {
    return new PDouble(this->value + std::any_cast<double>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PDouble::minus(Primitive* primitive) {
    return new PDouble(this->value - std::any_cast<double>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PDouble::times(Primitive* primitive) {
    return new PDouble(this->value * std::any_cast<double>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PDouble::div(Primitive* primitive) {
    return new PDouble(this->value / std::any_cast<double>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PDouble::rem(Primitive* primitive) {
    return new PDouble(std::fmod(this->value, std::any_cast<double>(primitive->getPrimitiveValue())));
}

Pulsar::Primitive* Pulsar::PDouble::compareGreater(Primitive* primitive) {
    return new PDouble(this->value > std::any_cast<double>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PDouble::compareLesser(Primitive* primitive) {
    return new PDouble(this->value < std::any_cast<double>(primitive->getPrimitiveValue()));
}

std::string Pulsar::PDouble::toString() {
    return std::to_string(this->value);
}
